### ISSUE TITLE
Example config for logstashCollector feature

### ISSUE THREAD

**Issue by Charlotte-Holt** [OpenLiberty/docs#1095]
Reach out to Don Bourne to determine SME for this. 

https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/twlp_analytics_logstash.html

**Comment by Rwalls1** [comment]
Link to current draft: https://draft-openlibertyio.mybluemix.net/docs/ref/feature/#logstashCollector-1.0.html

**Comment by Rwalls1** [comment]
@donbourne Please review: https://draft-openlibertyio.mybluemix.net/docs/ref/feature/#logstashCollector-1.0.html

**Comment by donbourne** [comment]
@Rwalls1 , 
not sure of the goals of the rewrite from original in the WebSphere Liberty KC (https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/twlp_analytics_logstash.html), but I think it was easier to follow the steps in the original than the new topic.

I would suggest taking the original topic and just making the following modifications to create the OL topic:
    - eliminate reference to Logstash V2.x, Elasticsearch V2.x, Kibana V4.x
    - eliminate reference to bluemixLogCollector
    - change "Enable HTTP access logging and use the '%h %u %t "%r" %s %b %D %{User-agent}i' log format." to "Enable HTTP access logging."

**Comment by Rwalls1** [comment]
@donbourne Are there specific pieces of that topic that you think should be included or adapted to avoid confusion?. The purpose of the modification was just to avoid including a full procedure as this topic is not intended for that, we just want to provide examples of how the feature can be used and explain why it would be used in certain contexts. We may eventually want to a link to a topic or guide that provides more detailed steps when one exists, which might help.

**Comment by donbourne** [comment]
If this topic is truly just going to be a config example, I would think we'd want to just provide the example and some description of why you might choose different settings for each of the configured values.  We would then still need another topic to cover the steps required to get it working since it's a lot more than just having the right config in the server.xml.

@Charlotte-Holt , are you guys thinking of having a "task" style topic for this one too?

**Comment by Rwalls1** [comment]
@donbourne Please review, I've updated the topic to only include the logstash configuration and the description of the parameters. What other details might be needed here if we're not including the task information? I will follow up with Charlotte to get her thoughts on having a separate task topic: https://draft-openlibertyio.mybluemix.net/docs/ref/feature/#logstashCollector-1.0.html

**Comment by donbourne** [comment]
let's meet to discuss this one -- I think the original WL topic on this does a better job of covering how to get up and going with logstashCollector since it details the steps beyond just setting the right config parameters.

**Comment by Rwalls1** [comment]
The topic information was deemed insufficient in the config example format and was consolidated into the following topic: https://github.com/OpenLiberty/docs/issues/1684

**Comment by Rwalls1** [comment]
@donbourne Please review the latest update: https://draft-openlibertyio.mybluemix.net/docs/ref/feature/#logstashCollector-1.0.html

**Comment by donbourne** [comment]
- should document that the garbageCollection source works only with IBM JDKs.  You might want to have a paragraph listing / describing the different sources in general.  (I think I saw comments elsewhere about us documenting the logstashCollector events -- so perhaps just a reference to that would do)

**Comment by Rwalls1** [comment]
@donbourne Can you provide a reference for the sources you are referring to?

**Comment by dmuelle** [comment]
Hi Richard- I'm writing the first draft now but have been unable top merge it today due to some issues with the Github branches & antora transition. The title of the doc will be `Logstash collector event fields` and the filename is `logstash-event-fields.adoc`. I'll post a link to the draft to this issue as soon as I'm able to publish it, hopefully tomorrow

**Comment by donbourne** [comment]
@Rwalls1 , by sources I just mean the messages, trace, audit, access log, (and garbage collection on ibm jvms).

**Comment by dmuelle** [comment]
Need to add another example to this doc per [LG-265/GH 1939](https://github.com/OpenLiberty/docs/issues/1939). 

**Replacing default  HTTP access log JSON fields**
In the Logstash Collector feature, you can use the `jsonAccessLogFields` attribute to replace the default JSON HTTP access log fields with a set of custom fields. You specify the replacement fields in the `logFormat` attribute of the `accesLogging` element within the `httpEndpoint element`. 

With this configuration, you can receive information that is otherwise not available in JSON logs, such as the remote user ID, request headers, and more. These logs can be used by log analysis tools, such as the Elastic stack, to better monitor your server.With this configuration, you can receive information that is otherwise not available in JSON logs, such as the remote user ID, request headers, and more. These logs can be used by log analysis tools, such as the Elastic stack, to better monitor your server.

The following example shows a configuration in the `server.xml` file to replace the default HTTP access log fields with custom fields that are specified by the `logFormat` attribute. xref:access-logging.adoc[HTTP access logging] must be enabled to receive JSON access logs.

    ```
    <httpEndpoint httpPort="9080" httpsPort="9443" id="defaultHttpEndpoint"> 
        <accessLogging logFormat='%R{W} %u %{my_cookie}C %s'/>
    </httpEndpoint>

    <logstashCollector 
        source=",accessLog" 
        jsonAccessLogFields="logFormat">
    </logstashCollector>
    ```

**Comment by Rwalls1** [comment]
@donbourne Please review: https://draft-openlibertyio.mybluemix.net/docs/20.0.0.8/reference/feature/logstashCollector-1.0.html#logstash-events-fields.adoc

**Comment by dmuelle** [comment]
@Rwalls1 sorry- the link I provided has the wrong syntax- the link to http access logging should be

 `xref:ROOT:access-logging.adoc[HTTP access logging]`

**Comment by Rwalls1** [comment]
@donbourne Please review the latest update: https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/reference/feature/logstashCollector-1.0.html

**Comment by donbourne** [comment]
Logstash collector gathers data from various sources and forwards the data to a Logstash server using Lumberjack protocol.
->
Logstash collector internally collects log data at run time and sends it directly to a Logstash server using Lumberjack protocol.

---

make this sentence be a separate paragraph on its own:
There are several parameters that you can use to configure the Logstash collector feature in the server.xml file.

---

remove:
Each log event type has customizable fields that help to manage the data that is sent to the configured Logstash log collection server. For more information, see Logstash collector event fields.

---

make this sentence be its own paragraph too:
The following example also has two tag elements in the logstashcollector element to add custom tags that decorate events within the application:

---

With this configuration, you can receive information that is otherwise not available in JSON logs, such as the remote user ID, request headers, and more. 
->
With this configuration, you can include fields such as the remote user ID, request headers, cookies and more in access log events. 

---

HTTP access logging must be enabled to receive JSON access logs.
->
HTTP access logging must be enabled for the Logstash collector feature to collect access log events.

**Comment by Rwalls1** [comment]
@donbourne Please review the latest update: https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/reference/feature/logstashCollector-1.0.html

**Comment by donbourne** [comment]
additional comments...

integer through 0 and 2147483647. ->  integer between 0 and 2147483647.

always capitalize "Elastic Stack"

source=",accessLog" -> source="accessLog"

**Comment by Rwalls1** [comment]
@lauracowen The introductory sentence "Logstash collector gathers data from various sources and forwards the data to a Logstash server using Lumberjack protocol." appears to be auto-generated. Can you help to update the sentence to align with Don's suggestion?

**Comment by lauracowen** [comment]
@Rwalls1 Good question. I've added a new section to the writing guidance here: https://ibm.box.com/s/90b7dnmkcu9qv4uple03niz6u3wj1np2 ("Getting the generated first sentence of the topic updated"). It's a while since I did it last so if you spot any problems in the instructions let me know and I'll update them. Ping me if you have any questions or anything's not clear.

**Comment by Rwalls1** [comment]
@donbourne Please review the latest update: https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/reference/feature/logstashCollector-1.0.html

**Comment by donbourne** [comment]
Replacing default HTTP access log JSON fields
->
Replacing default HTTP access log fields

---
the default JSON HTTP access log fields
->
the default set of HTTP access log fields

---
remove:
With this configuration, you can receive information that is otherwise not available in JSON logs, such as the remote user ID, request headers, and more. These logs can be used by log analysis tools, such as the Elastic Stack, to better monitor your server.

**Comment by Rwalls1** [comment]
@donbourne Please review the latest update: https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/reference/feature/logstashCollector-1.0.html

**Comment by Rwalls1** [comment]
@donbourne Please review the latest update: https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/reference/feature/logstashCollector-1.0.html

**Comment by donbourne** [comment]
looks good!

**Comment by Rwalls1** [comment]
@lauracowen Please review: https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/reference/feature/logstashCollector-1.0.html

**Comment by lauracowen** [comment]
Good start and nicely written but needs to be edited down to match the style of these topics:

"Logstash Collector configuration" example:
- Is this the minimal example required for this feature? If not, what would be and could you insert it as a separate example before this one? If it is, just call it "Basic configuration" (or whatever terminology we've used elsewhere). The whole page is about "Logstash Collector configuration", not just this example.
- Remove any superfluous config lines from the example - the aim isn't to provide a full-ish server.xml but to give the key lines that are required for the example. So the `featureManager` lines can be removed because how to enable the feature is described at the top of the page.
- These topics aren't providing how-to or detailed reference documentation for the configuration; they're simply providing example configuration and explanations of what that configuration would do if used in your own server.xml. So the first three paragraphs of introduction, whilst a nice introduction to using the feature, aren't relevant here - you're not trying to explain what is possible, but what this example shows. It needs to be far more minimal and focused. Take a look at some of the other similar topics. Briefly introduce the example, then give the example, then describe what the important values in the example would do if used (eg "the events are decorated with `serverRacka5` and `billingAppTeam` tags").
- If the information that's currently explained in the first three paragraphs of the example (eg about valid values, max lengths, etc) is not sufficiently explained in https://draft-openlibertyio.mybluemix.net/docs/latest/reference/config/logstashCollector.html (which is automatically linked further down in the topic), please can you raise that with the relevant developer to get that generated doc updated?
- Don't mention `server.xml` - it's not necessary as all this is in server.xml config.

"Replacing default HTTP access log fields" example:
- The first three paragraphs of this example explain the what and why of this example. Can you edit it down a bit into a single shorter paragraph that doesn't mention all the attributes? Just make it an introduction to what you can do without saying how at this stage. Then give the example. Then, after the example (or before it - we don't do this consistently across topics yet), describe what the values in that example do, which the current fourth paragraph does.

**Comment by Rwalls1** [comment]
@lauracowen Please review the latest update: https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/reference/feature/logstashCollector-1.0.html

**Comment by lauracowen** [comment]
Thanks. This is looking much more concise now.

- I'm still not sure that the first example is actually a basic configuration though. The attributes look like optional changes to the values (eg maxEvents) and the description says "you can change", implying it's not necessary and that the defaults are probably fine. Can you check with the SME what lines can be removed from the config to make it as simple and basic as possible?
- Also, the tags sound as if they're optional extras too and might be better in a second example snippet and not shown at all in the first unless a basic configuration is useless without them.
- In the second example, remove "use the jsonAccessLogFields attribute to" - it's not necessary and it's clear how you do it later in the example - important in the first sentence or so is what and why.
- Can you expand on what the actual example shows in terms of "the default HTTP access log fields with custom fields that are specified by the logFormat attribute"? eg give a translation of the custom fields or at least some of them as examples, based on the table that's in the HTTP access logging topic.
- Maybe also mention that the full logFormat syntax is available at that HTTP access logging topic?

**Comment by Rwalls1** [comment]
@donbourne Is there a more basic configuration for this feature?

**Comment by donbourne** [comment]
@Rwalls1 , the minimal config would be:

```xml
<logstashCollector
    hostName="localhost"
    port="5043"
    sslRef="mySSLConfig"
/>
```

that would just enable the "message" source.

**Comment by Rwalls1** [comment]
@donbourne Is the "message" source the default source for the feature or should there be an additional line that specifies the source?

**Comment by donbourne** [comment]
"message" is the default value for source.

**Comment by Rwalls1** [comment]
@lauracowen Please review: https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/reference/feature/logstashCollector-1.0.html

**Comment by lauracowen** [comment]
Thanks for getting a more minimal example. It's looking better now.

- [x] Can you remove the sentence about `message` source and, instead, alter the first sentence to say "...to enable the `message` source for..." or maybe better as "...to enable a source for data collection, which is, by default, message." (BTW, all the default values are listed if you click the `logstashCollector` link in the "Feature configuration elements" section lower down the topic.)
- [x] Then end the first sentence with a colon as lead-in to the example that follows it.
- [x] Can you move the closing `/>` to the end of the mySSLConfig line rather than being on a line of its own.
- [x] Can you tidy up the indentation of the second example which has more spaces on the nested lines than in the other two examples.
- [x] Okay, can we now split out the second example into two examples that try to do less at once (a given user won't necessarily want to do both things so it's easier to read the examples if they're focused on just one thing at a time)?
   - One example showing that you can change the max characters and max events (as the intro to the example states); also say that the logs are gather from six sources and explain what the `source` values mean.
   - Then the other example with its own heading about adding custom tags to decorate events, just show the minimal config with tags added and explain what these tags give you (eg "annotates the events to show that they came from serverRackA5 and billingAppTeam systems" or similar).
- [x] In the HTTP access example section, remove from the first sentence "that are specified by the logFormat attribute." as we don't need to know how initially, just what/why.
- [x] The sentence that says "The attribute specifies the `%{R}W...` log format options" is just describing what the reader can see if they look at the config snippet below - it doesn't add any value. If it's important for the user to understand this, the description needs to say what that means. Or, if that would be long-winded to describe just here, instead say that "The `logFormat` attribute specifies the log format options for the custom fields, which are listed in link:access-logging.adoc#_http_access_log_format[HTTP access logging]." (or whatever the correct link syntax is to link directly to the relevant section that actually lists the values). (and remove the "For more info.." sentence.


@donbourne Shouldn't the generated page of config settings (https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/reference/config/logstashCollector.html) list the valid values for `source`? It just gives the default of `message` in the table but doesn't explain what that means or what other sources would be valid and how to specify them?

**Comment by donbourne** [comment]
> Shouldn't the generated page of config settings (https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/reference/config/logstashCollector.html) list the valid values for source? It just gives the default of message in the table but doesn't explain what that means or what other sources would be valid and how to specify them?

yes -- that's a good point @lauracowen .  I've opened https://github.com/OpenLiberty/open-liberty/issues/13814 to get that fixed.

**Comment by Rwalls1** [comment]
@lauracowen Please review the latest update: https://draft-openlibertyio.mybluemix.net/docs/20.0.0.10/reference/feature/logstashCollector-1.0.html

**Comment by Yushan-Lin** [comment]
> Shouldn't the generated page of config settings (https://draft-openlibertyio.mybluemix.net/docs/20.0.0.9/reference/config/logstashCollector.html) list the valid values for source? It just gives the default of message in the table but doesn't explain what that means or what other sources would be valid and how to specify them?

@lauracowen I opened an issue with the updated description https://github.com/OpenLiberty/docs/issues/2581

**Comment by lauracowen** [comment]
Thanks @Rwalls1, looking good.

The config examples vary between examples in their XML syntax. Some have a closing tag while others have the closing slash within the opening tag:
- In the second example, is there a reason for using a closing XML tag instead of just `/>`? The first example uses `/>`. If it's valid both ways and it's just down to preference/formatting, that's fine, I think (eg in the fourth example, the `logstashCollector` element might look odd and harder to read without a proper closing tag?).
- In the third example, is it valid XML to have the `/>` straight after a nested element's closing tag? (ie `</tag> />`)? It doesn't look right. (But also the closing tag should be on the previous line anyway, as you did in the previous example.)

Might be a question for @donbourne  or @Yushan-Lin ?


Other than that, a minor style thing:
- Can you make sure the lead-in sentence before an example ends with a colon (to introduce the example)? In the fourth example, the sentence about "The following example" is not immediately before the example itself so I can see why it doesn't. If it would be weird to put a colon on the end of the second sentence, just move the second sentence (which explains the example) to below the example, then have the "The following example..." sentence be the lead-in that introduces the example with a colon? It's fine to have the explanation of the example beneath the example, as long as the text above the example introduces the purpose/why of the example.

**Comment by donbourne** [comment]
I think 2nd example should be:

```xml
<logstashCollector
    source="message,trace,garbageCollection,ffdc,accessLog,audit"
    hostName="localhost"
    port="5043"
    sslRef="mySSLConfig"
    maxEvents="10"
    maxFieldLength="3000" />
```

3rd example should be:
```xml
<logstashCollector
    hostName="localhost"
    port="5043"
    sslRef="mySSLConfig">
    <tag>serverRackA5</tag>
    <tag>billingAppTeam</tag>
</logstashCollector>
```

4th example should be:
```xml
<httpEndpoint httpPort="9080" httpsPort="9443" id="defaultHttpEndpoint">
    <accessLogging logFormat='%R{W} %u %{my_cookie}C %s'/>
</httpEndpoint>
<logstashCollector
    source="accessLog"
    jsonAccessLogFields="logFormat" />
```

**Comment by Rwalls1** [comment]
@lauracowen Thanks, please review the latest update: https://draft-openlibertyio.mybluemix.net/docs/20.0.0.10/reference/feature/logstashCollector-1.0.html

**Comment by lauracowen** [comment]
Great thanks, Richard. Signing off.

**Comment by dmuelle** [comment]
# Peer review

examples look good, just a few minor comments

## Basic configuration

I think the intro sentence here is a little confusing. It says
`The following example shows how to configure the Logstash Collector feature to enable a source for data collection, which is, by default, message:`
But then the example actually shows a hostname, port number and ssl ref. Those are the minimum required config for the feature to work, but the intro only mentions `source`, which is set by default here and not visible in the example. I think it would be easier to understand if the intro talked about what is in the example, in addition to explaining the default source. Something like 
`To use the Logstash collector feature, at minimum, you must specify the hostname and port number of a Logstash server and the ID of your SSL configuration. The following example shows how to configure the Logstash Collector feature to collect data from the default source, which is message:`

## Customizing Logstash Collector

- This heading doesn't tell you much about the example, and could apply to any of these examples. To give more context, you could change it to something like `Setting the maximum message length and number of events`


- `You can also limit the maximum number of events that are sent for each type of event per second.`

maybe

`You can also limit the number of events that are sent per second for each event type.` 

- Also wondering if you need to say ` ...in the logstashCollector element` since that is the only config element for this feature and all the examples on this page are w/in that element. I don't think you need to mention it for any of these examples.

## Adding custom event tags

see previous comment re the logstashCollector element

## Replacing default HTTP access log fields

`...replace the default HTTP access log fields with custom fields`

I think it's a little confusing to refer to these fields as "custom", since OL JSON logging has [a different feature that enables users to set custom JSON fields](https://draft-openlibertyio.mybluemix.net/docs/20.0.0.11/log-trace-configuration.html#fields). 

The replacement fields themselves are not "custom", but a configured set of alternate fields specified by `logFormat`. In [the Log and trace configuration doc](https://draft-openlibertyio.mybluemix.net/docs/20.0.0.11/log-trace-configuration.html#http) they're referred to as a "different set of fields".

**Comment by Rwalls1** [comment]
@dmuelle Thanks, I have addressed your feedback.

**Comment by Rwalls1** [comment]
New link: https://draft-openlibertyio.mybluemix.net/docs/20.0.0.12/reference/feature/logstashCollector-1.0.html

**Comment by Rwalls1** [comment]
Editing reference: https://learning.oreilly.com/library/view/developing-quality-technical/9780133119046/ch06.html
I used this chapter as a guide to ensure that the description of the config examples were clear and avoided ambiguity. The tips related to avoiding wordiness also helped me to review the descriptions to ensure conciseness.

**Comment by dmuelle** [comment]
Doc is now on vNext branch, will publish with 20.0.0.12 on 11.20

**Comment by chirp1** [comment]
Hi Richard, another nice config snippet! I have a few comments:
- Change "...from the default source, which is message:" to "from the default source, which is the message source". (I find the original wording confusing.) Also, add highlighting to "message".
- I think you unnecessarily separated "You can also change the maximum number of characters for the stack trace in an FFDC event" from "...that are allowed for the message field in a message or trace event." and attached it to a separate thought of "limit the number of events that are sent per second for each event type.". How about if you reword the info to something like  "...that are allowed for the message field in a message, trace, or FFDC event. You can limit the number of events that are sent per second for each event type."
- How about for clarity if you work into the example that the max 3000 characters and the max 10 events applies to the message, trace, accessLog, ffdc, garbageCollection, and audit events.
-  Although I see the phrase "decorate events within the application" in the topic that you drew from, I don't know what it means. How about if you go to the "logstashCollector" link farther down in your Open Libety topic and use the description for the tag name to reword the sentence that contains "decorate". 
- For the phrase "The logFormat attribute specifies..." I don't think that in this example "logFormat" is an attribute even though it is an attribute for the HTTP access log format. Look at the info for jsonAccessLogFields at the logstashCollector link and reword as appropriate.

**Comment by Rwalls1** [comment]
@chirp1 I've addressed your feedback: https://draft-openlibertyio.mybluemix.net/docs/21.0.0.8/reference/feature/logstashCollector-1.0.html
The `logFormat` attribute sentence refers to the attribute specified by the `accessLogging` element. I've added another sentence to describe the `logFormat` value that is specified by the `jsonAccessLogFields` attribute, which I think you were referring to in your initial review feedback.

**Comment by chirp1** [comment]
Updates look good! Moving to Ready to Publish.

**Comment by Rwalls1** [comment]
Staging: https://docs-staging-openlibertyio.mybluemix.net/docs/21.0.0.8/reference/feature/logstashCollector-1.0.html

**Comment by Rwalls1** [comment]
Updates have been merged to Vnext and will publish with the 21.0.0.8 release.
